src/components/registration/steps.jsx

```diff
@@ -82,38 +82,53 @@ module.exports = {
         onChangeShowPassword: function (field, value) {
             this.setState({showPassword: value});
         },
-        onValidSubmit: function (formData, reset, invalidate) {
-            this.setState({waiting: true});
+        validateUsername: function (username) {
             api({
                 host: '',
-                uri: '/accounts/check_username/' + formData.user.username + '/'
+                uri: '/accounts/check_username/' + username + '/'
             }, function (err, res) {
                 var formatMessage = this.props.intl.formatMessage;
-                this.setState({waiting: false});
-                if (err) return invalidate({all: err});
+                if (err) {
+                    this.refs.form.refs.formsy.updateInputsWithError({all: err});
+                    return false;
+                }
                 res = res[0];
                 switch (res.msg) {
                 case 'valid username':
                     this.setState({
                         validUsername: 'pass'
                     });
-                    return this.props.onNextStep(formData);
+                    return true;
                 case 'username exists':
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameExists'})
                     });
+                    return false;
                 case 'bad username':
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameVulgar'})
                     });
+                    return false;
                 case 'invalid username':
                 default:
-                    return invalidate({
+                    this.refs.form.refs.formsy.updateInputsWithError({
                         'user.username': formatMessage({id: 'registration.validationUsernameInvalid'})
                     });
+                    return false;
```

Comment:
once I add the callback (which I forgot to do), `true/false` handles the difference between `onValidSubmit` and `onSubmit`. Since we don't want to call `onNextStep` for `onSubmit`, but do for `onValidSubmit`.